### PR TITLE
case-lib: is_ipc4: use new debugfs file instead of kernel parameter

### PR DESCRIPTION
**Pull request by plbossart**
First step to remove use of kernel parameters

Link: https://github.com/thesofproject/linux/pull/4902
Link: https://github.com/thesofproject/sof-test/issues/1166

### CODE REVIEW

**Comment by plbossart** [comment]
shellcheck error is a pre-existing issue that was obviously overlooked

````
In case-lib/lib.sh line 954:
        TGLU_RVP_NOCODEC_IPC4ZPH | ADLP_RVP_NOCODEC_IPC4ZPH | MTLP_RVP_NOCODEC |ADLP_RVP_NOCODEC-ipc4 | TGLU_RVP_NOCODEC-ipc4 | MTLP_RVP_NOCODEC | MTLP_RVP_NOCODEC-multicore-2cores | MTLP_RVP_NOCODEC-multicore-3cores | LNLM_RVP_NOCODEC)
                                                              ^--------------^ SC22[21](https://github.com/thesofproject/sof-test/actions/runs/8544615585/job/23411051491?pr=1167#step:3:22) (warning): This pattern always overrides a later one on line 954.
                                                                                                                                ^--------------^ SC[22](https://github.com/thesofproject/sof-test/actions/runs/8544615585/job/23411051491?pr=1167#step:3:23)22 (warning): This pattern never matches because of a previous pattern on line 954.
````

MTLP_RVP_NOCODEC  is listed twice in the same OR.

File: case-lib/lib.sh [hunk 1]
```diff
@@ -810,9 +810,31 @@ is_firmware_file_zephyr()
...
+}
+
 is_ipc4()
 {
     local ipc_type
+
+    # First try with debugfs
+    if is_ipc4_debugfs; then
+	return 0
+    fi
```
Comment (marc-hb) [inline case-lib/lib.sh:835]:
```shell
if ipc_type=$(cat /sys/kernel/debug/sof/ipc_type); then
   # "cat" was successful
   case $ipc_type in
   0) return ... # in shell, success = error code 0 = "true"
   1) return ...
   *) die "unexpected /sys/kernel/debug/sof/ipc_type=$ipc_type"
fi

# "cat" failed, fallback on ...
```

**Review by marc-hb — CHANGES_REQUESTED** [review]
cc:
- https://github.com/thesofproject/linux/issues/3867

```
    # If /sys/module/snd_sof_pci/parameters/ipc_type does not exist
    # the DUT is running IPC3 mode
```
This shortcut and bad assumption should print a big fat `dlogw()` TODAY and be "upgraded" to a `die()` failure in the future. Please add the `dlogw()` now; I think that would have saved you time and it will save others'

File: case-lib/lib.sh [hunk 2]
```diff
@@ -810,9 +810,31 @@ is_firmware_file_zephyr()
...
 }
 
+is_ipc4_debugfs()
+{
+    local ipc_type
+
+    ipc_file=/sys/kernel/debug/sof/ipc_type
+    ipc_type=$(cat $ipc_file) || {
+        return 1
+    }
```
[the remark this answers is not in the document]
Comment (plbossart) [reply case-lib/lib.sh:820]: the overloading is already present in the initial code for is_ipc4().

And while I am at it it's just torture to see code a branch " if is_ipc4" taken when ipc4 is NOT used. 
Such mind-twisting is just not right. someone will screw-up when the value of 'true' is inverted.

File: case-lib/lib.sh [hunk 3]
```diff
@@ -810,9 +810,31 @@ is_firmware_file_zephyr()
...
+
+    if [ "$ipc_type" -eq 1 ]; then
+        return 0
+    fi
+    return 1
```
[the remark this answers is not in the document]
Comment (plbossart) [reply case-lib/lib.sh:825]: Or keep the syntax simple for the rest of us who just understand if then...

The simplest way to write a branch is the best... 

I vote to disallow any 'smart' bash construct which requires people like me to consult an online help. bash is a means to an end.

File: case-lib/lib.sh (hunk 1 above)
Comment (plbossart) [reply case-lib/lib.sh:835]: Good suggestion that anyone would understand, thanks.

**Comment by marc-hb** [comment]
> MTLP_RVP_NOCODEC is listed twice in the same OR.

Fix submitted in #1168

File: case-lib/lib.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (marc-hb) [reply case-lib/lib.sh:820]: > the overloading is already present in the initial code for is_ipc4().

It's not because the higher level does something questionable that the lower level, internal implementation must align with it. Of course it shouldn't be different without any good reason either but there's no "strict" requirement there.

File: case-lib/lib.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (marc-hb) [reply case-lib/lib.sh:825]: I must differ here sorry.

I totally agree that bash is quirky and the confusion between booleans and error codes plays a large role. It is a real pain. However, learning about that particular confusion between booleans and error codes is unfortunately not optional. especially not in sof-TEST code, the main purpose of any test code being to deal with _errors_.

For instance, until a person can make sense of what I wrote in #312, then they're not allowed to have an informed opinion on this topic! I think that boolean/error code knowledge is really the minimum required. Note I just wrote "make sense" of #312, _not_ "remember it all".

So now back to this particular example: wrapping `test_something` into `if test_something; then return 0; else return 1` is:
- "useful" only for people who MISS this basic boolean/error code knowledge.
- very verbose because you'd have to do this wrapping pretty much everywhere
- distracting because it's exposing the fact that 0 = true instead of abstracting it away.


All this being said, a one-line comment never hurts. For example:

```shell
is_something()
{
   # stuff

   # return true if X exists
   test -e X
}
```

[the remark this answers is not in the document]
Comment (plbossart) [reply case-lib/lib.sh:825]: You're entitled to your opinions.

I personally don't know *anything* to bash and everything I read about it gives me a migraine. I won't remember any of this next Monday anyways.

so if I can use if/else I will...

File: case-lib/lib.sh [hunk 4]
```diff
@@ -813,22 +813,47 @@ is_firmware_file_zephyr()
...
-    # -1: DUT runs IPC3 mode, is_ipc4 return 1(false)
-    # 1: DUT runs IPC4 mode, is_ipc4 return 0(true)
-    if [ "$ipc_type" -eq 1 ]; then
-        return 0
+    # "cat" failed, backwards-compatible fallback with ipc_type kernel parameter
+    if ipc_type=$(cat /sys/module/snd_sof_pci/parameters/ipc_type); then
+	# "cat" was successful
+	case $ipc_type in
+	    -1)
+		# the parameter is not set, assume IPC3 was used.
```
Comment (marc-hb) [inline case-lib/lib.sh:843]: Can you please "upgrade" this comment to a `dlogw`
```shell
dlogw "snd_sof_pci/parameters/ipc_type parameter is not set, assume IPC3 was used."
```

File: case-lib/lib.sh [hunk 5]
```diff
@@ -813,22 +813,47 @@ is_firmware_file_zephyr()
...
-    # the DUT is running IPC3 mode
-    ipc_type=$(cat $ipc_file) || {
-        return 1
-    }
+    if ipc_type=$(cat /sys/kernel/debug/sof/fw_profile/ipc_type); then
+	# "cat" was successful
+	case $ipc_type in
+	    0)
+		return 1
+		;;
```
Comment (marc-hb) [inline case-lib/lib.sh:828]: Nit: this is a lot of whitespace...

```shell
  0) return 1 ;;
```

File: case-lib/lib.sh [hunk 6]
```diff
@@ -813,22 +813,47 @@ is_firmware_file_zephyr()
...
+		return 0 # in shell, success = error code 0 = "true"
+		;;
+	    *)
+		die "invalid ipc_type in /sys/module/snd_sof_pci/parameters/ipc_type=$ipc_type"
+		;;
+	esac
     fi
-    return 1
+
+    # If /sys/module/snd_sof_pci/parameters/ipc_type does not exist
```
Comment (marc-hb) [inline case-lib/lib.sh:860]: upgrade this comment to a `dlogw` too.

File: case-lib/lib.sh [hunk 7]
```diff
@@ -296,9 +296,15 @@ fake_kern_error()
...
-        local fw_path
-        fw_path=$(cat /sys/module/snd_sof_pci/parameters/fw_path)
+    local fw_path
+    local fw_path_info='/sys/kernel/debug/sof/fw_path'
+
+    if test ! -e $fw_path_info; then
+	fw_path_info='/sys/module/snd_sof_pci/parameters/fw_path'
+    fi
+
+    if fw_path=$(cat $fw_path_info); then
```
Comment (marc-hb) [inline case-lib/lib.sh:306]: This commit adds `if ... ; then` but no `fi` termination so I don't see how this can work?

File: case-lib/lib.sh [hunk 8]
```diff
@@ -296,9 +296,15 @@ fake_kern_error()
...
 {
     local subdir='intel/sof' # default
-    if test -e /sys/module/snd_sof_pci/parameters/fw_path; then
-        local fw_path
-        fw_path=$(cat /sys/module/snd_sof_pci/parameters/fw_path)
+    local fw_path
+    local fw_path_info='/sys/kernel/debug/sof/fw_path'
+
+    if test ! -e $fw_path_info; then
+	fw_path_info='/sys/module/snd_sof_pci/parameters/fw_path'
```
Comment (marc-hb) [inline case-lib/lib.sh:303]: Nit: the shorter, "idiomatic" style is:
```shell
test -e $fw_path_info ||
    fw_path_info='/sys/module/snd_sof_pci/parameters/fw_path'
```

It reads in plain English likes this: "either $fw_path_info exists, OR we redefine $fw_path_info and point it at the alternative"

**Review by marc-hb — COMMENTED** [review]
Looks good except for some nits and one... missing `fi`?

File: case-lib/lib.sh [hunk 9]
```diff
@@ -813,22 +819,47 @@ is_firmware_file_zephyr()
...
+	esac
+    fi
 
-    # If /sys/module/snd_sof_pci/parameters/ipc_type exists
-    # If the value of file ipc_type is:
-    # -1: DUT runs IPC3 mode, is_ipc4 return 1(false)
-    # 1: DUT runs IPC4 mode, is_ipc4 return 0(true)
-    if [ "$ipc_type" -eq 1 ]; then
-        return 0
+    # "cat" failed, backwards-compatible fallback with ipc_type kernel parameter
```
Comment (marc-hb) [inline case-lib/lib.sh:838]: Some vague indication in the comment of which kernel version will be required would be useful.

File: case-lib/lib.sh (hunk 7 above)
Comment (marc-hb) [reply case-lib/lib.sh:306]: Never mind, now I see what happened.

File: case-lib/lib.sh (hunk 4 above)
Comment (plbossart) [reply case-lib/lib.sh:843]: done

File: case-lib/lib.sh (hunk 8 above)
Comment (plbossart) [reply case-lib/lib.sh:303]: done

File: case-lib/lib.sh (hunk 9 above)
Comment (plbossart) [reply case-lib/lib.sh:838]: done

**Review by marc-hb — APPROVED** [review]

**Comment by marc-hb** [comment]
Lots of models missing but everything green for the ones present

**Review by kv2019i — APPROVED** [review]

**Comment by marc-hb** [comment]
This does not work actually: it's at least missing a few `sudo`. I think you didn't notice because you run sof-test as root but that's not how it's usually tested.

https://sof-ci.01.org/softestpr/PR1167/build265/devicetest/index.html?model=MTLP_RVP_NOCODEC&testcase=verify-kernel-module-load-probe

```
2024-04-05 14:58:59 UTC [REMOTE_INFO] Starting func_exit_handler(0)
cat: /sys/kernel/debug/sof/fw_profile/ipc_type: Permission denied
2024-04-05 14:58:59 UTC [REMOTE_WARNING] using pre kernel 6.9 backwards-compatible fallback with ipc_type kernel parameter
cat: /sys/kernel/debug/sof/fw_profile/ipc_type: Permission denied
2024-04-05 14:58:59 UTC [REMOTE_WARNING] using pre kernel 6.9 backwards-compatible fallback with ipc_type kernel parameter
2024-04-05 14:58:59 UTC [REMOTE_INFO] pkill -TERM -f mtrace-reader.py
2024-04-05 14:58:59 UTC [REMOTE_INFO] ktime=123 sof-test PID=3129: ending
2024-04-05 14:58:59 UTC [REMOTE_INFO] Test Result: PASS!
```

**Comment by marc-hb** [comment]
Filed new issue:
- #1171
